File: runtime/native/osx/src/TiMenuFactory.m
/**
 * This file is part of Appcelerator's Titanium project.
 *
 * Copyright 2008 Appcelerator, Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License. 
 */

#import "TiMenuFactory.h"

@implementation TiMenuFactory

- (id)initWithWebView:(WebView *)wv 
{
	self = [super init];
	if (self != nil) 
	{
		webView = wv; // assign only. don't retain. prevents retain loop memory leak
	}
	return self;
}

- (void)dealloc
{
	webView = nil;
	[super dealloc];
}

- (TiUserMenu *)createUserMenu:(NSString*)label
{
	//TODO: review with marshall - probably need a getUserMenu
	NSMenu *appMenu = [[[NSApp mainMenu] itemWithTitle:label] submenu];
	if (appMenu == nil)
	{
		//FIXME
	}
	else
	{
		return [[TiUserMenu alloc]initWithMenu:appMenu];
	}
	return nil;
}


- (TiSystemMenu *)createSystemMenu:(NSString*)iconURL caption:(NSString*)caption callback:(WebScriptObject*)callback;
{
	return [[TiSystemMenu alloc] initWithURL:iconURL caption:caption callback:callback];
}

#pragma mark -
#pragma mark WebScripting

+ (BOOL)isSelectorExcludedFromWebScript:(SEL)sel {
	return (nil == [self webScriptNameForSelector:sel]);
}


+ (NSString *)webScriptNameForSelector:(SEL)sel 
{
	if (sel == @selector(createSystemMenu:caption:callback:)) {
		return @"createSystemMenu";
	}
	else if (sel == @selector(createUserMenu:)) {
		return @"createUserMenu";
	}
	return nil;
}


+ (BOOL)isKeyExcludedFromWebScript:(const char*)key {
	return (nil == [self webScriptNameForKey:key]);
}


+ (NSString *)webScriptNameForKey:(const char *)name {
	return nil;
}

@end

File: runtime/native/win32/ti_shell/src/ti_menu_factory.cc
/*
* Copyright 2006-2008 Appcelerator, Inc.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

#include "ti_menu_factory.h"
#include "ti_system_menu.h"

TiMenuFactory::TiMenuFactory()
{
	BindMethod("createSystemMenu", &TiMenuFactory::createSystemMenu);
}

void TiMenuFactory::createSystemMenu(const CppArgumentList& args, CppVariant* result)
{
	if (args.size() >= 3) {
		if (args[0].isString() && args[1].isString() && args[2].isObject()) {
			std::string iconURL = args[0].ToString();
			std::string caption = args[1].ToString();
			NPVariant variant;
			args[2].CopyToNPVariant(&variant);

			NPObject* callback = NPVARIANT_TO_OBJECT(variant);

			TiSystemMenu *menu = new TiSystemMenu(iconURL, caption, callback);

			result->Set(menu->ToNPObject());
		}
	}
}

void TiMenuFactory::createUserMenu(const CppArgumentList& args, CppVariant* result)
{

}